---
title: "Two subtle problems with over-representation analysis"
date: "`r Sys.Date()`"
output:
  rmdformats::robobook:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 5
    fig_height: 5
bibliography: references.bib
csl: plos-computational-biology.csl
---

[Anusuiya Bora<sup>1,2</sup>](https://orcid.org/0009-0006-2908-1352),
[Mark Ziemann<sup>1,2</sup>*](https://orcid.org/0000-0002-7688-6974)

**Affiliations**

1. Burnet Institute, Melbourne, Australia.

2. Deakin University, School of Life and Environmental Sciences, Geelong, Australia.

(*) Corresponding author: mark.ziemann@burnet.edu.au

# Abstract

TODO.

# Keywords

* Bioinformatics

* Pathway analysis

* Gene set enrichment analysis

* Research integrity

# Main Body

## Introduction

```{r,libs,echo=FALSE}

library("DiagrammeR")
library("RColorBrewer")
library("kableExtra")
#library("gt")

```

Functional enrichment analysis involves the summarisation of omics data to reflect
biological differences, and has become one of the most popular methods in bioinformatics.
Enrichment analysis software packages have been collectively cited 131,332 times as of
late 2019 [@Xie2021-ch].
The algorithms used by these tools can largely be divided into two groups, functional
class scoring (FCS) and over-representation analysis (ORA).
In FCS, all detected genes are ranked by their degree of differential expression followed
by a statistical test for enrichment of gene categories at either extreme of the
ranked list.
ORA involdes the selection of genes of interest, followed by a test to ascertain whether
certain functional categories are over-represented in the selected genes.

According to a study that used simulated differential expression data, FCS accuracy was
superior to ORA over a range of experimental designs [@Kaspi2020-rn].
Despite this performance difference, a survey of articles published in 2019 showed that
seven of the top eight most popular tools were based on ORA [@Wijesooriya2022-fw].
The discrepancy between performance and popularity likely arose due to the relative ease of
conducting ORA as compared to FCS.
A minimalist ORA might involve pasting a single list of genes into a text box on a website,
where the results appear nearly instantaneously.
On the other hand, FCS typically involves downloading specific softwares and dealing with a
dataset that represents every detected gene, typically ~20,000 rows and ensuring that the
format from upstream tools can be shoehorned to meet the requirements of FCS packages.

Since it's initial development in 1999 [@Tavazoie1999-bu], ORA has proliferated widely,
becoming part of many websites and software packages.
The most popular ORA website is DAVID [@Dennis2003-ar;@Huang2009-gi;@Sherman2022-jn],
followed by Ingenuity Pathway Analysis, a commercial software package provided by QIAGEN Inc.
ORA has been implemented into popular R/Bioconductor packages including `ClusterProfiler`
[@Yu2012-jj],`limma` [@Ritchie2015-oz] and `GOseq` [@Young2010-iw].

Despite the popularity of ORA, there are concerns that this technique is being misused.
Due to technological and biological reasons, not all genes are detected in transcriptomic
studies, meaning that some genes will more readily appear as differentially expressed.
To address this bias, a background list of detected genes is required for comparison to the
list of selected genes (foreground list) [@Tilford2009-vx].
Failure to use a background gene list can lead to dramatic changes in enrichment results,
rendering them unreliable [@Timmons2015-ki].
Unfortunately, use of an appropriate background list is reported in only a small fraction
(~4%) of peer-reviewed articles describing enrichment analysis [@Wijesooriya2022-fw].
In addition, correction of p-values for multiple testing is crucial in controlling the
false positive rate, but this is reported in only ~54% of studies [@Wijesooriya2022-fw].
These discoveries have resulted in the development of best practices for end users
to improve awareness of the issues and help enforce reporting standards [@Zhao2023-mx].

The focus of this work is to raise awareness to two existing problems in ORA as implemented
by some popular enrichment tools.
The first problem is that genes belonging to the background list are removed from the analysis
entirely if they are not annotated as belonging to any functional categories.
This results in the background list appearing smaller than it should, leading to an
underestimation of fold enrichment scores and significance values.
In principle, this problem would differentially affect analyses involving different gene set
libraries, with smaller libraries such as Kyoto Encyclopedia of Genes and Genomes (KEGG)
[@Kanehisa2023-ni], which describes 2,727 genes affected more severely as compared to larger
libraries like Gene Ontology [@The_Gene_Ontology_Consortium2023-xp] which describes 19,428
genes (figures obtained from MSigDB [@Liberzon2015-cs]).

The second problem is that adjusting p-values for multiple testing, also known as
false-discovery correction is implemented improperly.
ORA tools should implement a threshold by which a pathway could be considered as detected.
Ideally, a pathway should be considered detected based on the presence of a predetermined
number of member genes in the background set.
However some tools use the foreground set to define a pathway as detected.
This is problematic, as the act of trying to find an intersect between a foreground gene list
and a pathway is in itself a test, even if there are no common genes.
This results in pathways being discarded from the analysis if no common genes are found in
the foreground, despite a large number of those genes being present in the background list.
This leads to FDR values appearing lower than they should.
We expect this problem to have a more severe effect when the foreground list is fairly small.

Here, we show the effect of these two problems on real RNA-seq based enrichment analysis
results and demonstrate these two issues lead to poor performance using simulated expression
data.
Finally we classify popular tools regarding these two problems and provide general
recommendations for end users.

## A systematic assessment of reproducibility in enrichment analysis

### Methods


### Results

## Discussion

# Data and Software Availability

## Underlying data

Publicly available data were obtained from Digital Expression Explorer 2 (dee2.io).

## Software and Code

* The code repository including template Dockerfile and R Markdown script are available on GitHub
(https://github.com/markziemann/enrichment_recipe).

* The example Docker image is available on DockerHub (https://hub.docker.com/r/mziemann/enrichment_recipe).

* The code repository and Docker image have been uploaded to Zenodo for long-term archiving
(https://zenodo.org/record/8170984).

## Availability of other materials

* The full protocol can be found at protocols.io
(https://www.protocols.io/view/a-recipe-for-extremely-reproducible-enrichment-ana-j8nlkwpdxl5r/v2) @Ziemann2023-kz.

* Instructional video guides are available on YouTube
(https://www.youtube.com/playlist?list=PLAAydBPtqFMXDpLa796q7f7W1HK4t_6Db).

## Author Contributions

Conceptualization: MZ, AB.
Data Curation: MZ.
Formal Analysis: MZ.
Funding Acquisition: N/A.
Investigation: MZ, AB.
Methodology: MZ, AB.
Project Administration: MZ.
Resources: MZ.
Software: MZ, AB.
Supervision: MZ.
Validation: AB.
Visualization: MZ, AB.
Writing – Original Draft Preparation: MZ, AB.
Writing – Review & Editing: MZ, AB.

## Competing Interests

No competing interests were disclosed.

## Grant Information

The authors declared that no grants were involved in supporting this work.

## Acknowledgements

We thank Ms Megan Soria and Ms Kaumadi Wijesooriya (Deakin University) for discussions on the concept and manuscript.
This research was supported by use of the Nectar Research Cloud, a collaborative Australian research
platform supported by the NCRIS-funded Australian Research Data Commons (ARDC).
The authors gratefully acknowledge the contribution to this work of the
Victorian Operational Infrastructure Support Program received by the Burnet Institute.

# Bibliography
